### PR TITLE
Write the NUnit arguements to a temp file and pass that to NUnit console

### PR DESCRIPTION
**Pull request by BlythMeister**
### Description

Another option for #2113 to always write NUnit args to a file and call nunit with the path to the args.

### CODE REVIEW

**Comment by BlythMeister** [comment]
Not sure why this or the other fail ..

**Comment by matthid** [comment]
I guess this (and the other one) wants pull from release/next as well

**Comment by BlythMeister** [comment]
Ok I'll sort that all tomorrow.  Laptop is off for the night now 👍

**Comment by BlythMeister** [comment]
Still don't get why this is failing...I included the latest...

File: src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs [hunk 1]
```diff
@@ -325,6 +329,9 @@ let run (setParams : NUnit3Params -> NUnit3Params) (assemblies : string seq) =
             FileName = tool
             WorkingDirectory = getWorkingDir parameters
             Arguments = args }) >> Process.withFramework) processTimeout
+
+    File.Delete(path)
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs]: maybe try-finally?

File: src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs [hunk 2]
```diff
@@ -316,7 +316,11 @@ let run (setParams : NUnit3Params -> NUnit3Params) (assemblies : string seq) =
     let assemblies = assemblies |> Seq.toArray
     if Array.isEmpty assemblies then failwith "NUnit: cannot run tests (the assembly list is empty)."
     let tool = parameters.ToolPath
-    let args = buildArgs parameters assemblies
+    let generatedArgs = buildArgs parameters assemblies
+    let path = Path.Combine(Path.GetTempPath(), (sprintf "%s.txt" (Guid.NewGuid().ToString())))
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs]: What about `Path.GetTempFileName()`?

**Comment by matthid** [comment]
nevermind the failures, mono is stupid as usually. Added a new test and after fixing it on windows it still fails on mono. Should work soon

File: src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs (hunk 1 above)
Comment (BlythMeister) [reply src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs]: Agreed, will update

File: src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs (hunk 2 above)
Comment (BlythMeister) [reply src/app/Fake.DotNet.Testing.NUnit/NUnit3.fs]: I didn't know that was a thing.  Will update (and update a lot of other code too 😂)

**Comment by BlythMeister** [comment]
@matthid i think this is just that build error you told me to ignore :)

**Comment by BlythMeister** [comment]
@matthid any chance of getting this into 5.8.2?

**Comment by matthid** [comment]
Honestly I want to get some other smaller fixes in as well, but I don't want to be held back into feature creep :) So I guess we probably should release something and then try to get the rest in and release again...

**Comment by matthid** [comment]
Could we add a small integration test, or would that add to much headache (I'd be fine with restricting the test to CIs where .NET or mono is available)?

**Comment by matthid** [comment]
> I'm not really sure how to do this.

I'd just add a new test to the "Fake.Core.IntegrationTests" project, add a project reference to `Fake.DotNet.Testing.NUnit` and run nunit with it (on some prepared project you can check-in)

**Comment by BlythMeister** [comment]
@matthid i started looking at this and got lost in the whole getting the nunit console runner in place.
Also, would this test not actually be testing the nunit runner...and not the fake implementation.

**Comment by matthid** [comment]
@BlythMeister Yes generally I feel like we need an easy way to mock the `Fake.Core.Process` API, this way we could add unit-tests for the cli generation. But I have not come up with a good way yet.

**Comment by BlythMeister** [comment]
@matthid would you be ok to accept this without that test...or do i need to work out how to get this functioning :(

**Comment by matthid** [comment]
@BlythMeister I'm ok if you say you tested it manually :) Problem is that new features like this might break people and having at least some rudimentary test-infrastructure in place makes it much easier for me and others to add regression tests if needed.

I'm just in the process of trying to improve azure devops integration into PRs (which is red at the moment), which might make it easier to deploy on staging from a PR to test stuff, but its a lot of overhead for stuff which I think could/should be tested more "locally"

**Comment by BlythMeister** [comment]
I've run this in a mock rig without full FAKE running (linqpad script) to make sure it calls nunit ok.

But i agree, there should be a nicer way to test that the arguments being run in a process are correct without actually needing the process...but that feels really hard with the way Fake.Core.Process works,.

**Comment by BlythMeister** [comment]
The other thing making this tricky is that the file created has a temporary name...and is deleted once finished...

**Comment by matthid** [comment]
@BlythMeister I think the following could work (will try to get some POC working locally later, maybe even with this PR):

Add another (internal) overload which gets some "process starter interface", this overload could be used for unit testing and the public one just calls this internal method with the default process api.

**Comment by BlythMeister** [comment]
@matthid i've never done that sort of thing in F# before, would be interested to see it in action if you could add to this PR :D

**Comment by matthid** [comment]
@BlythMeister It would look like https://github.com/fsharp/FAKE/pull/2131

**Comment by BlythMeister** [comment]
Ooo nice, does that pr mean this one is redundant?  Both can go in together?

**Comment by BlythMeister** [comment]
@matthid the stuff over there looks good.  I think the test you added does show the expected behaviour.
Granted not with all the arugument flags but the logic to generate args has not changed.

**Pull request merged**